docs/cudf/source/developer_guide/ARCHITECTURE.md

```diff
@@ -0,0 +1,199 @@
+cuDF does not support the arbitrary `object` dtype, but instead defines a few custom types for common use-cases:
+- `ListDtype`: Lists where each element in every list in a Column is of the same type.
+- `StructDtype`: Dicts where a given key always maps to values of the same type
+- `CategoricalDtype`: Analogous to the pandas categorical dtype except that the categories are stored in device memory.
+- `DecimalDtype`: Fixed-point numbers
+- `IntervalDtype`: Intervals
+
+Note that there is a many-to-one mapping between data type and `Column` class.
+For instance, all numerical types (floats and ints of different widths) are all managed using `NumericalColumn`.
+
+
+### Buffer
+
+Although a `Column` represents an Arrow-compliant data structure, it does not directly handle memory mangaement.
+That job is delegated to the `Buffer` class, which represents a device memory allocation that it _may or may not_ own.
+A `Buffer` constructed from a preexisting device memory allocation (such as a CuPy array) will view that memory.
+Conversely, a `Buffer` constructed from a host object will allocate new device memory and copy in the data.
+cuDF uses the [RMM](https://github.com/rapidsai/rmm) library for allocating device memory.
+You can read more about device memory allocation with RMM [here](https://github.com/rapidsai/rmm#devicebuffers).
+
+```{note}
+cuDF needs to interoperate with a wide range of Python libraries, many of which allocate device memory.
+The ownership behavior described above is designed to minimize new memory allocations for externally-owned memory.
+cuDF's use of `libcudf`, however, represents a slight break from the paradigm employed above:
+since `libcudf` objects are not Python objects, we need something to manage their lifetimes from within cuDF.
+
+The key is to recognize that `libcudf` offers both owning (e.g. `column`) and non-owning (e.g. `column_view`) objects.
+All `libcudf` algorithms accept views as parameters while returning (new) owning objects.
+When calling `libcudf` APIs, cuDF Python construct views from cudf `Buffers`.
+When owning objects are returned, cuDF has an rmm object take ownership of that memory and stores that in a `Buffer`.
+The result is that all memory allocated by `libcudf` inside cuDF eventually has ownership transferred to a `Buffer`.
+```
+
+## The Cython layer
+
+The lowest level of cuDF is its interaction with `libcudf` via Cython.
+Most algorithms in cudf follow a similar pattern.
+The `Frame` layer processes inputs and calls a `Column` method.
+That method in turn does some additional processing before finally calling a Cython function.
+The result is then passed back up through the layers, undergoing postprocessing as needed.
+
+The Cython layer itself is largely composed of two parts: C++ bindings and Cython wrappers.
+We use Cython to expose C++ functionality to Python.
+This code essentially consists of copying libcudf header files into new files with a slightly different format.
```

Comment:
```suggestion
The C++ binding code declares the contents of libcudf header files (`.hpp`) in [Cython declaration files (`.pxd`)](https://cython.readthedocs.io/en/latest/src/tutorial/pxd_files.html).
```

Reply:
I rewrote these few sentences to try and make this section a bit clearer. The sentences around this were all a bit muddy.